modules/azure-public-dns/dns-records.tf

```diff
@@ -0,0 +1,173 @@
+# Since the azurerm provider uses a separate resource type for each DNS record
+# type, we'll need to split up our input list.
+locals {
+  recordsets       = { for rs in var.recordsets : rs.type => rs... }
+  a_recordsets     = lookup(local.recordsets, "A", [])
+  aaaa_recordsets  = lookup(local.recordsets, "AAAA", [])
+  cname_recordsets = lookup(local.recordsets, "CNAME", [])
+  mx_recordsets    = lookup(local.recordsets, "MX", [])
+  ns_recordsets    = lookup(local.recordsets, "NS", [])
+  ptr_recordsets   = lookup(local.recordsets, "PTR", [])
+  srv_recordsets   = lookup(local.recordsets, "SRV", [])
+  txt_recordsets   = lookup(local.recordsets, "TXT", [])
+
+  # Some of the resources only deal with one record at a time, and so we need
+  # to flatten these.
+  cname_records = flatten([
+    for rs in local.cname_recordsets : [
+      for r in rs.records : {
+        name = rs.name
+        type = rs.type
+        ttl  = rs.ttl
+        data = r
+      }
+    ]
+  ])
+
+  # Next two expressions will raise an exception if an unsupported type is entered
+  # instead of silently ignoring it
+  supported_record_types = {
+    A     = true
+    AAAA  = true
+    CNAME = true
+    MX    = true
+    NS    = true
+    PTR   = true
+    SRV   = true
+    TXT   = true
+  }
+
+  check_supported_types = [
+    # The index operation here will fail if one of the records has
+    # an unsupported type.
+    for rs in var.recordsets : local.supported_record_types[rs.type]
+  ]
+}
```

Comment:
can we separate locals from resource blocks? 
The idea is to have locals in the component and sent to the module. The only thing I want to focus on this repo is consistency. Since locals contain list of variables that are different for each environment so that can be passed as an argument to the module.